File: axiom/neuron/viterbi.php
<?php
require_once '..\..\utility.php';
$i = 0;
$txt[$i++] = "\\({s}_{t} = \\sum\\limits_{i=1}^{t} {G}_{{y}_{i},{y}_{i - 1}} + \\sum\\limits_{i=0}^{t} {x}_{i,{y}_{i}}\\tag*{Eq.s_definition}\\)";
$txt[$i++] = "\\({x'}_{t} = [{y}_{t}]\\min\\limits_{{y}_{0:t}} {s}_{t}\\tag*{Eq.x_quote_definition}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = {x}_{t + 1} + \\min \\left({x'}_{t} + G\\right)\\tag*{Eq.recursion}\\)";
$txt[$i++] = "\\(\\min\\limits_{{y}_{0:t + 1}} {s}_{t} = \\min {x'}_{t}\\tag*{Eq.aggregate}\\)";
$txt[$i++] = "\\({s}_{t + 1} = \\sum\\limits_{i=1}^{t + 1} {G}_{{y}_{i},{y}_{i - 1}} + \\sum\\limits_{i=0}^{t + 1} {x}_{i,{y}_{i}}\\tag*{Eq[0]}\\)";
$txt[$i++] = "\\({s}_{t + 1} - {s}_{t} = - \\sum\\limits_{i=1}^{t} {G}_{{y}_{i},{y}_{i - 1}} + \\sum\\limits_{i=1}^{t + 1} {G}_{{y}_{i},{y}_{i - 1}} - \\sum\\limits_{i=0}^{t} {x}_{i,{y}_{i}} + \\sum\\limits_{i=0}^{t + 1} {x}_{i,{y}_{i}}\\tag*{Eq[1]}\\)";
$txt[$i++] = "\\({s}_{t + 1} = {G}_{{y}_{t + 1},{y}_{t}} + {s}_{t} + {x}_{t + 1,{y}_{t + 1}}\\tag*{Eq[2]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = [{y}_{t + 1}]\\min\\limits_{{y}_{0:t + 1}} {s}_{t + 1}\\tag*{Eq[3]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = [{y}_{t + 1}]\\min\\limits_{{y}_{0:t + 1}} \\left({G}_{{y}_{t + 1},{y}_{t}} + {s}_{t} + {x}_{t + 1,{y}_{t + 1}}\\right)\\tag*{Eq[4]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = {x}_{t + 1} + [{y}_{t + 1}]\\min\\limits_{{y}_{0:t + 1}} \\left({G}_{{y}_{t + 1},{y}_{t}} + {s}_{t}\\right)\\tag*{Eq[5]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = {x}_{t + 1} + [{y}_{t + 1}]\\min\\limits_{{y}_{t}} \\left({G}_{{y}_{t + 1},{y}_{t}} + \\min\\limits_{{y}_{0:t}} {s}_{t}\\right)\\tag*{Eq[6]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = {x}_{t + 1} + [{y}_{t + 1}]\\min \\left({G}_{{y}_{t + 1}} + [{y}_{t}]\\min\\limits_{{y}_{0:t}} {s}_{t}\\right)\\tag*{Eq[7]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = {x}_{t + 1} + \\min \\left(G + [{y}_{t}]\\min\\limits_{{y}_{0:t}} {s}_{t}\\right)\\tag*{Eq[8]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = {x}_{t + 1} + \\min \\left({x'}_{t} + G\\right)\\tag*{Eq.recursion}\\)";
$txt[$i++] = "\\(\\min\\limits_{{y}_{t}} \\min\\limits_{{y}_{0:t}} {s}_{t} = \\min {x'}_{t}\\tag*{Eq[9]}\\)";
$txt[$i++] = "\\(\\min [{y}_{t}]\\min\\limits_{{y}_{0:t}} {s}_{t} = \\min {x'}_{t}\\tag*{Eq[10]}\\)";
$txt[$i++] = "\\(\\text{True}\\)";
render(__FILE__, $txt);
?>        

File: axiom/neuron/viterbi.py

from sympy.core.symbol import Symbol

from sympy.utility import plausible
from sympy.core.relational import Equality

from sympy.core.numbers import oo
from sympy.concrete.expr_with_limits import Ref, MIN
from sympy.concrete.summations import Sum


@plausible
def apply(G, x, y):
    _, d = x.shape
    i = Symbol('i', integer=True)
    t = Symbol('t', integer=True, nonnegative=True)

    s = Symbol('s', shape=(oo,),
                    definition=Ref[t](Sum[i:1:t](G[y[i], y[i - 1]]) + Sum[i:0:t](x[i, y[i]])))

    x_quote = Symbol("x'", shape=(oo, d), definition=Ref[y[t], t](MIN[y[0:t]](s[t])))

    return Equality(x_quote[t + 1], x[t + 1] + MIN(x_quote[t] + G)), \
        Equality(MIN[y[:t + 1]](s[t]), MIN(x_quote[t]))


from sympy.utility import check


@check
def prove(Eq):
    d = Symbol('d', integer=True)

    # oo is the length of the sequence
    # d is the number of output labels
    G = Symbol('G', shape=(d, d), real=True)
    x = Symbol('x', shape=(oo, d), real=True)
    y = Symbol('y', shape=(oo,), real=True)

    Eq.s_definition, Eq.x_quote_definition, Eq.recursion, Eq.aggregate = apply(G, x, y)
    
    Eq.x_quote_definition = Eq.x_quote_definition.reference((Eq.x_quote_definition.lhs.indices[-1],))
    
    s = Eq.x_quote_definition.rhs.function.function.base

    t = Eq.recursion.lhs.indices[0] - 1
    Eq << Eq.s_definition.subs(t, t + 1)

    Eq << Eq[-1] - Eq.s_definition

    Eq << Eq[-1].this.rhs.simplify() + s[t]

    Eq << Eq.x_quote_definition.subs(t, t + 1)

    Eq << Eq[-1].this.rhs.subs(Eq[-2])

    Eq << Eq[-1].this.rhs.function.simplify()

    Eq << Eq[-1].this.rhs.args[1].function.bisect(back=1)

    Eq << Eq[-1].this.rhs.args[1].function.as_Ref()

    Eq << Eq[-1].this.rhs.args[1].as_Min()

    Eq << Eq[-1].this.rhs.subs(Eq.x_quote_definition.reversed)

    Eq << Eq.aggregate.this.lhs.bisect(back=1)

    Eq << Eq[-1].this.lhs.as_Ref()

    Eq << Eq[-1].subs(Eq.x_quote_definition.reversed)


if __name__ == '__main__':
    prove(__file__)
